core/src/main/java/org/apache/calcite/statistic/QuerySqlStatisticProvider.java

```diff
@@ -0,0 +1,192 @@
+import org.apache.calcite.plan.RelOptSchema;
+import org.apache.calcite.plan.RelOptTable;
+import org.apache.calcite.plan.ViewExpanders;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.rel2sql.RelToSqlConverter;
+import org.apache.calcite.rel.rel2sql.SqlImplementor;
+import org.apache.calcite.sql.SqlDialect;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.tools.Frameworks;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.Util;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import javax.sql.DataSource;
+
+/**
+ * Implementation of {@link SqlStatisticProvider} that generates and executes
+ * SQL queries.
+ */
+public class QuerySqlStatisticProvider implements SqlStatisticProvider {
+  private final Consumer<String> sqlConsumer;
+
+  /** Creates a QuerySqlStatisticProvider.
+   *
+   * @param sqlConsumer Called when each SQL statement is generated
+   */
+  public QuerySqlStatisticProvider(Consumer<String> sqlConsumer) {
+    this.sqlConsumer = Objects.requireNonNull(sqlConsumer);
+  }
+
+  public double tableCardinality(RelOptTable table) {
+    final JdbcTable jdbcTable = table.unwrap(JdbcTable.class);
+    return withBuilder(jdbcTable.jdbcSchema,
+        (cluster, relOptSchema, jdbcSchema, relBuilder) -> {
+          // Generate:
+          //   SELECT COUNT(*) FROM `EMP`
+          relBuilder.push(table.toRel(ViewExpanders.simpleContext(cluster)))
+              .aggregate(relBuilder.groupKey(), relBuilder.count());
+
+          final String sql = toSql(relBuilder.build(), jdbcSchema.dialect);
+          final DataSource dataSource = jdbcSchema.getDataSource();
+          try (Connection connection = dataSource.getConnection();
+               Statement statement = connection.createStatement();
+               ResultSet resultSet = statement.executeQuery(sql)) {
+            if (!resultSet.next()) {
+              throw new AssertionError();
+            }
+            return resultSet.getDouble(1);
+          } catch (SQLException e) {
+            throw handle(e, sql);
+          }
+        });
+  }
+
+  public boolean isForeignKey(RelOptTable fromTable, List<Integer> fromColumns,
+      RelOptTable toTable, List<Integer> toColumns) {
+    final JdbcTable jdbcTable = fromTable.unwrap(JdbcTable.class);
+    return withBuilder(jdbcTable.jdbcSchema,
+        (cluster, relOptSchema, jdbcSchema, relBuilder) -> {
+          // EMP(DEPTNO) is a foreign key to DEPT(DEPTNO) if the following
+          // query returns 0:
+          //
+          //   SELECT COUNT(*) FROM (
+          //     SELECT deptno FROM `EMP` WHERE deptno IS NOT NULL
+          //     MINUS
+          //     SELECT deptno FROM `DEPT`)
```

Comment:
We also need to check that the toColumn are primary key of toTable

Reply:
By design, the `isForeignKey` method only checks referential integrity - i.e. that there is at least one matching row. It is left to `isKey` to determine "at most one".